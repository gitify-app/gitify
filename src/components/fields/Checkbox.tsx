import type { FC, ReactNode } from 'react';
import { Tooltip } from './Tooltip';

export interface ICheckbox {
  name: string;
  label: string;
  helpText?: ReactNode | string;
  tooltip?: ReactNode | string;
  checked: boolean;
  disabled?: boolean;
  onChange: (evt: React.ChangeEvent<HTMLInputElement>) => void;
}

export const Checkbox: FC<ICheckbox> = (props: ICheckbox) => {
  return (
    <div className="mb-3 mt-1 text-sm">
      <div className="flex items-start">
        <div className="flex h-5 items-center">
          <input
            type="checkbox"
            id={props.name}
            className="size-4 rounded border-gray-300 text-indigo-600 focus:ring-indigo-500"
            checked={props.checked}
            onChange={props.onChange}
            disabled={props.disabled}
          />
        </div>

        <div className="ml-3">
          <label
            htmlFor={props.name}
            className="font-medium text-gray-700 dark:text-gray-200"
            style={
              props.disabled ? { textDecoration: 'line-through' } : undefined
            }
          >
            {props.label}
            {props.tooltip && (
              <Tooltip name={`tooltip-${props.name}`} tooltip={props.tooltip} />
            )}
          </label>
        </div>
      </div>

      {props.helpText && (
        <div className="mt-1 text-xs italic text-gray-500 dark:text-gray-300">
          {props.helpText}
        </div>
      )}
    </div>
  );
};
