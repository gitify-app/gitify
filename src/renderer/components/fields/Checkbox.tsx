import type { FC, ReactNode } from 'react';

import { CounterLabel, Stack } from '@primer/react';

import { cn } from '../../utils/cn';
import { Tooltip } from './Tooltip';

export interface ICheckbox {
  name: string;
  label: string;
  counter?: number;
  tooltip?: ReactNode | string;
  checked: boolean;
  disabled?: boolean;
  visible?: boolean;
  onChange: (evt: React.ChangeEvent<HTMLInputElement>) => void;
}

export const Checkbox: FC<ICheckbox> = ({
  visible = true,
  ...props
}: ICheckbox) => {
  const counter = props.counter > 0 ? props.counter : undefined;

  return (
    visible && (
      <Stack
        direction="horizontal"
        gap="condensed"
        align="center"
        className="text-sm"
      >
        <input
          type="checkbox"
          id={props.name}
          className="size-4 rounded-sm cursor-pointer"
          checked={props.checked}
          onChange={props.onChange}
          disabled={props.disabled}
          data-testid={`checkbox-${props.name}`}
        />

        <label
          htmlFor={props.name}
          className={cn(
            'font-medium text-gitify-font cursor-pointer',
            props.disabled && 'line-through',
          )}
        >
          {props.label}
        </label>

        {props.tooltip && (
          <Tooltip name={`tooltip-${props.name}`} tooltip={props.tooltip} />
        )}

        {counter && (
          <CounterLabel scheme={props.checked ? 'primary' : 'secondary'}>
            {counter}
          </CounterLabel>
        )}
      </Stack>
    )
  );
};
