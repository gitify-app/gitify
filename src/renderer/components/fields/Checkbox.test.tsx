import { render } from '@testing-library/react';
import { Checkbox, type ICheckbox } from './Checkbox';

describe('renderer/components/fields/Checkbox.tsx', () => {
  const props: ICheckbox = {
    name: 'appearance',
    label: 'Appearance',
    checked: true,
    onChange: jest.fn(),
  };

  it('should render - visible', () => {
    const tree = render(<Checkbox {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render - not visible', () => {
    const tree = render(<Checkbox visible={false} {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render - disabled', () => {
    const tree = render(<Checkbox disabled={true} {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render - tooltip', () => {
    const tree = render(
      <Checkbox {...props} tooltip={<div>Hello world</div>} />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render - positive counter unselected', () => {
    const tree = render(<Checkbox {...props} counter={5} checked={false} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render - positive counter selected', () => {
    const tree = render(<Checkbox {...props} counter={5} checked={true} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render - zero counter', () => {
    const tree = render(<Checkbox {...props} counter={0} />);
    expect(tree).toMatchSnapshot();
  });
});
