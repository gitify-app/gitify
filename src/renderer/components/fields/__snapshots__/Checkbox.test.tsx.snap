// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer/components/fields/Checkbox.tsx should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mt-3 mb-2 text-sm"
      >
        <div
          class="flex items-center"
        >
          <input
            checked=""
            class="size-4 rounded-sm cursor-pointer"
            id="appearance"
            type="checkbox"
          />
          <div
            class="flex items-center ml-3"
          >
            <label
              class="font-medium text-gitify-font cursor-pointer"
              for="appearance"
            >
              Appearance
            </label>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mt-3 mb-2 text-sm"
    >
      <div
        class="flex items-center"
      >
        <input
          checked=""
          class="size-4 rounded-sm cursor-pointer"
          id="appearance"
          type="checkbox"
        />
        <div
          class="flex items-center ml-3"
        >
          <label
            class="font-medium text-gitify-font cursor-pointer"
            for="appearance"
          >
            Appearance
          </label>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
